### PR TITLE
Add support for binding axis_name in gmap

### PR DESCRIPTION
**Pull request by apaszke**
This allows executing collectives over the gmapped axes. This requires
some extra manipulation of the gmapped jaxpr, since gmap exposes a
single logical axis name, but evaluates the program using multiple
"physical" axes.

This also fixes some bugs around handling `multiple_returns` in
vmap collective implementation.

### CODE REVIEW

File: jax/experimental/general_map.py [hunk 1]
```diff
@@ -125,3 +151,19 @@ def sched_fun_wrapper(*args):
...
+def subst_axis_names(jaxpr, replaced_name, axis_names):
+  eqns = [subst_eqn_axis_names(eqn, replaced_name, axis_names) for eqn in jaxpr.eqns]
+  return core.Jaxpr(jaxpr.constvars, jaxpr.invars, jaxpr.outvars, eqns)
+
+def subst_eqn_axis_names(eqn, replaced_name, axis_names):
+  if isinstance(eqn.primitive, (core.CallPrimitive, core.MapPrimitive)):
+    if eqn.params.get('axis_name', None) == replaced_name:  # Check for shadowing
+      return eqn
+    new_call_jaxpr = subst_axis_names(eqn.params['call_jaxpr'], replaced_name, axis_names)
+    return eqn._replace(params=dict(eqn.params, call_jaxpr=new_call_jaxpr))
```
Comment (mattjj) [inline jax/experimental/general_map.py:165]: Is `_replace` defined anywhere?

File: jax/core.py [hunk 2]
```diff
@@ -1576,7 +1576,7 @@ def extend_axis_env(axis_name, size: int, tag: Any):
       yield
     finally:
       frame_ = thread_local_state.trace_state.axis_env.pop()
-      assert frame is frame_
+    assert frame is frame_  # Only runs if there was was no exception
```
Comment (mattjj) [inline jax/core.py:1579]: Hmm don't we want to run this check even if there is an exception, to ensure that our cleanup procedure is happening as expected? O maybe that obscures the underlying exception or something?

File: jax/interpreters/batching.py [hunk 3]
```diff
@@ -146,7 +146,9 @@ def process_primitive(self, primitive, tracers, params):
         if len(axis_names) > 1:
           return split_axis(primitive, axis_name, tracers, params)
         vals_out, dims_out = collective_rules[primitive](vals_in, dims_in, frame.size, **params)
-        return map(partial(BatchTracer, self), vals_out, dims_out)
+        results = map(partial(BatchTracer, self), vals_out, dims_out)
+        print(results)
```
Comment (mattjj) [inline jax/interpreters/batching.py:150]:
```suggestion
```

File: tests/gmap_test.py [hunk 4]
```diff
@@ -57,12 +77,30 @@ def f(x):
...
+          "vmap collectives only supported when omnistaging is enabled")
+  def testAxisName(self, schedule):
+    def f(x):
+      return x - lax.psum(x, 'i')
+    x = jnp.arange(8)
+    self.assertAllClose(gmap(f, schedule, axis_name='i')(x),
+                        vmap(f, axis_name='i')(x))
+
+  @ignore_gmap_warning()
+  @skipIf(not config.omnistaging_enabled,
```
Comment (mattjj) [inline tests/gmap_test.py:95]: Just to double-check, are you able to run these tests by setting the omnistaging env var or similar? I ask because if flag parsing happens after module import time (i.e. after this expression is evaluated) it might not be possible to enable these tests.

**Review by mattjj — APPROVED** [review]
LGTM! A few questions noted below.

File: jax/core.py (hunk 2 above)
Comment (apaszke) [reply jax/core.py:1579]: Yeah it has obscured another problem making it really hard to debug.

File: jax/experimental/general_map.py (hunk 1 above)
Comment (apaszke) [reply jax/experimental/general_map.py:165]: https://docs.python.org/3.8/library/collections.html#collections.somenamedtuple._replace

File: tests/gmap_test.py (hunk 4 above)
Comment (apaszke) [reply tests/gmap_test.py:95]: yeah those tests run if I prefix my `pytest` invocation with `JAX_OMNISTAGING=1`
